providers/decoders.inc

```diff
@@ -7,36 +7,77 @@
  * https://www.openssl.org/source/license.html
  */
 
-#ifndef DECODER
-# error Macro DECODER undefined
+#ifndef DECODER_PROVIDER
+# error Macro DECODER_PROVIDER undefined
 #endif
 
+#define DECODER_STRUCTURE_type_specific_keypair         "type-specific"
+#define DECODER_STRUCTURE_type_specific_params          "type-specific"
+#define DECODER_STRUCTURE_type_specific                 "type-specific"
+#define DECODER_STRUCTURE_type_specific_no_pub          "type-specific"
+#define DECODER_STRUCTURE_PKCS8                         "pkcs8"
+#define DECODER_STRUCTURE_SubjectPublicKeyInfo          "SubjectPublicKeyInfo"
+#define DECODER_STRUCTURE_DH                            "dh"
+#define DECODER_STRUCTURE_DHX                           "dhx"
+#define DECODER_STRUCTURE_DSA                           "dsa"
+#define DECODER_STRUCTURE_EC                            "ec"
+#define DECODER_STRUCTURE_RSA                           "rsa"
+
+#define DECODER(name, input, output, fips)                              \
```

Comment:
It has been pointed out quite a few times now,
that not using underscores for these macros breaks the build of a specific platform.
i.e it cannot deal with fips=" #fips

Reply:
Sorry, that point flew past me.  Okie, will fix

Reply:
I've added a comment on this, to help other who might wonder why that underscore was added.
*For emphasis: we **should** add such comments rather than trusting the illusion that everyone has followed every discussion and remember them all*